plugins/methodinfo/src/main/java/org/qbicc/plugin/methodinfo/MethodData.java

```diff
@@ -0,0 +1,136 @@
+package org.qbicc.plugin.methodinfo;
+
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+final class MethodData {
+    private final IndexableSet<String> fileNameSet = new IndexableSet<>();
+    private final IndexableSet<String> classNameSet = new IndexableSet<>();
+    private final IndexableSet<String> methodNameSet = new IndexableSet<>();
+    private final IndexableSet<String> methodDescSet = new IndexableSet<>();
+    private final IndexableSet<MethodInfo> methodInfoSet = new IndexableSet<>();
+    private final IndexableSet<SourceCodeInfo> sourceCodeInfoSet = new IndexableSet<>();
+    private InstructionMap[] instructionMapList;
+
+    MethodData(int instructionsCount) {
+        this.instructionMapList = new InstructionMap[instructionsCount];
+    }
+
+    int add(SourceCodeInfo sourceCodeInfo) {
+        return sourceCodeInfoSet.addIfAbsent(sourceCodeInfo);
+    }
+
+    int add(MethodInfo methodInfo) {
+        return methodInfoSet.addIfAbsent(methodInfo);
+    }
+
+    void add(int index, InstructionMap instructionMap) {
+        instructionMapList[index] = instructionMap;
+    }
+
+    MethodInfo[] getMethodInfoTable() {
+        return methodInfoSet.toArray(new MethodInfo[0]);
+    }
+
+    SourceCodeInfo[] getSourceCodeInfoTable() {
+        return sourceCodeInfoSet.toArray(new SourceCodeInfo[0]);
+    }
+
+    InstructionMap[] getInstructionMapList() {
+        InstructionMap[] map = new InstructionMap[instructionMapList.length];
+        System.arraycopy(instructionMapList, 0, map, 0, map.length);
+        return map;
+    }
+
+    int addFileName(String fileName) {
+        return fileNameSet.addIfAbsent(fileName);
+    }
+
+    String[] getFileNameList() { return fileNameSet.toArray(new String[0]); }
+
+    int addClassName(String className) {
+        return classNameSet.addIfAbsent(className);
+    }
+
+    String[] getClassNameList() { return classNameSet.toArray(new String[0]); }
+
+    int addMethodName(String methodName) {
+        return methodNameSet.addIfAbsent(methodName);
+    }
+
+    String[] getMethodNameList() { return methodNameSet.toArray(new String[0]); }
+
+    int addMethodDesc(String methodDesc) {
+        return methodDescSet.addIfAbsent(methodDesc);
+    }
+
+    String[] getMethodDescList() { return methodDescSet.toArray(new String[0]); }
+
+    /**
+     * A set implementation that allows to retrieve index of the items in the order they are added to the set.
+     * @param <T>
+     */
+    static final class IndexableSet<T> extends AbstractSet<T> implements Set<T> {
```

Comment:
`LinkedHashSet` wouldn't serve the purpose as I need to get the index of the elements as they get added to the set. `LinkedHashSet` preserves the order but does not provide the index, which is why I am using a map internally to store index corresponding to each element.
However, I realized use of `ConcurrentHashMap` is not correct as it would not preserve the order of elements when they are retrieved - in this case using `toArray`. I have changed it to `LinkedHashMap`. Also I think I don't need it to support concurrent retrieval and updates, but it does need synchronization when adding the elements, so making `add` and `addIfAbsent` as synchronized should be sufficient.

Reply:
> Also I think I don't need it to support concurrent retrieval and updates, but it does need synchronization when adding the elements, so making `add` and `addIfAbsent` as synchronized should be sufficient.

That's a difficult invariant to maintain.  If the `add` operations need to be synchronized, then so do the consumers api points

Reply:
I could have tried `Collections.synchronizedSet` but then it would not support `addIfAbsent` operation on the set. I think I would have to make all methods as synchronized to mimic `Collections.synchronizedSet`.